modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,256 @@
+    impl->predict(*this, frame.getMat(), outputs);
+}
+
+ClassificationModel::ClassificationModel(const String& model, const String& config)
+    : Model(model, config) {};
+
+ClassificationModel::ClassificationModel(const Net& network) : Model(network) {};
+
+std::pair<int, float> ClassificationModel::classify(InputArray frame)
+{
+    std::vector<Mat> outs;
+    impl->predict(*this, frame.getMat(), outs);
+    CV_Assert(outs.size() == 1);
+
+    double conf;
+    cv::Point maxLoc;
+    minMaxLoc(outs[0].reshape(1, 1), nullptr, &conf, nullptr, &maxLoc);
+    return {maxLoc.x, static_cast<float>(conf)};
+}
+
+void ClassificationModel::classify(InputArray frame, int& classId, float& conf)
+{
+    std::tie(classId, conf) = classify(frame);
+}
+
+DetectionModel::DetectionModel(const String& model, const String& config)
+    : Model(model, config) {};
+
+DetectionModel::DetectionModel(const Net& network) : Model(network) {};
+
+void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
+                   CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect>& boxes,
+                   float confThreshold, float nmsThreshold)
+{
+    std::vector<Mat> detections;
+    impl->predict(*this, frame.getMat(), detections);
+
+    boxes.clear();
+    confidences.clear();
+    classIds.clear();
+
+    int frameWidth  = frame.cols();
+    int frameHeight = frame.rows();
+    if (getLayer(0)->outputNameToIndex("im_info") != -1)
+    {
+        frameWidth = impl->size.width;
+        frameHeight = impl->size.height;
+    }
+
+    std::vector<String> layerNames = getLayerNames();
+    int lastLayerId = getLayerId(layerNames.back());
+    Ptr<Layer> lastLayer = getLayer(lastLayerId);
+
+    std::vector<int> predClassIds;
+    std::vector<Rect> predBoxes;
+    std::vector<float> predConf;
+    if (lastLayer->type == "DetectionOutput")
+    {
+        // Network produces output blob with a shape 1x1xNx7 where N is a number of
+        // detections and an every detection is a vector of values
+        // [batchId, classId, confidence, left, top, right, bottom]
+        for (int i = 0; i < detections.size(); ++i)
+        {
+            float* data = (float*)detections[i].data;
+            for (int j = 0; j < detections[i].total(); j += 7)
+            {
+                float conf = data[j + 2];
+                if (conf < confThreshold)
+                    continue;
+
+                int left   = data[j + 3];
+                int top    = data[j + 4];
+                int right  = data[j + 5];
+                int bottom = data[j + 6];
+                int width  = right  - left + 1;
+                int height = bottom - top + 1;
+
+                if (width * height <= 1)
+                {
+                    left   = data[j + 3] * frameWidth;
+                    top    = data[j + 4] * frameHeight;
+                    right  = data[j + 5] * frameWidth;
+                    bottom = data[j + 6] * frameHeight;
```

Comment:
@l-bat, in https://github.com/opencv/opencv/issues/14841 we figured out that it's more robust to multiply predicted values from DetectionOutput by `frameWidth - 1` and `frameHeight - 1` because it produces values in range [0, 1].